File: php/features.php
<?php
declare(strict_types=1);

// BranchQuickLinks SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class BranchQuickLinksFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new BranchQuickLinksBaseFeature();
            case "debug":
                return new BranchQuickLinksDebugFeature();
            case "idempotency":
                return new BranchQuickLinksIdempotencyFeature();
            case "metrics":
                return new BranchQuickLinksMetricsFeature();
            case "paging":
                return new BranchQuickLinksPagingFeature();
            case "ratelimit":
                return new BranchQuickLinksRatelimitFeature();
            case "retry":
                return new BranchQuickLinksRetryFeature();
            case "test":
                return new BranchQuickLinksTestFeature();
            case "timeout":
                return new BranchQuickLinksTimeoutFeature();
            default:
                return new BranchQuickLinksBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/branchquicklinks_sdk/features.py
# BranchQuickLinks SDK feature factory

from branchquicklinks_sdk.feature.base_feature import BranchQuickLinksBaseFeature
from branchquicklinks_sdk.feature.debug_feature import BranchQuickLinksDebugFeature
from branchquicklinks_sdk.feature.idempotency_feature import BranchQuickLinksIdempotencyFeature
from branchquicklinks_sdk.feature.metrics_feature import BranchQuickLinksMetricsFeature
from branchquicklinks_sdk.feature.paging_feature import BranchQuickLinksPagingFeature
from branchquicklinks_sdk.feature.ratelimit_feature import BranchQuickLinksRatelimitFeature
from branchquicklinks_sdk.feature.retry_feature import BranchQuickLinksRetryFeature
from branchquicklinks_sdk.feature.test_feature import BranchQuickLinksTestFeature
from branchquicklinks_sdk.feature.timeout_feature import BranchQuickLinksTimeoutFeature


_FEATURES = {
    "base": lambda: BranchQuickLinksBaseFeature(),
    "debug": lambda: BranchQuickLinksDebugFeature(),
    "idempotency": lambda: BranchQuickLinksIdempotencyFeature(),
    "metrics": lambda: BranchQuickLinksMetricsFeature(),
    "paging": lambda: BranchQuickLinksPagingFeature(),
    "ratelimit": lambda: BranchQuickLinksRatelimitFeature(),
    "retry": lambda: BranchQuickLinksRetryFeature(),
    "test": lambda: BranchQuickLinksTestFeature(),
    "timeout": lambda: BranchQuickLinksTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
